vms/subnet.go

```diff
@@ -0,0 +1,257 @@
+package vms
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/ava-labs/avalanche-network-runner/network"
+
+	"github.com/ava-labs/avalanchego/api"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/constants"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/vms/platformvm"
+)
+
+// CustomVM wraps data to create a custom VM
+type CustomVM struct {
+	Path     string
+	Genesis  string
+	Name     string
+	SubnetID string
+	ID       string
+}
+
+const (
+	loopTimeout         = 1 * time.Second
+	longTimeout         = 10 * loopTimeout
+	defaultKeyThreshold = 1
+
+	validatorWeight    = 3000
+	validatorStartDiff = 30 * time.Second
+	validatorEndDiff   = 30 * 24 * time.Hour // 30 days
+)
+
+// SetupSubnet creates the necessary transactions to create a subnet for a given custom VM.
+// It then also waits until all these transactions are confirmed on all nodes.
+// Finally it creates a blockchain and makes sure all nodes of the given network
+// are validating this blockchain.
+// It requires a `privateKey` in order to issue the necessary transactions
+func SetupSubnet(
+	ctx context.Context,
+	log logging.Logger,
+	vm CustomVM,
+	network network.Network,
+	privateKey string,
+) error {
+	log.Info("creating subnet")
+
+	userPass := api.UserPass{
+		Username: "test",
+		Password: "vmsrkewl",
+	}
+
+	allNodes, err := network.GetAllNodes()
+	if err != nil {
+		return err
+	}
+
+	txNodeNames, err := network.GetNodeNames()
+	if err != nil {
+		return err
+	}
+	// txNode will be the node we issue all transactions on
+	txNode := allNodes[txNodeNames[0]]
+	txClient := txNode.GetAPIClient()
+	// we need to create a user for platformvm calls
+	ok, err := txClient.KeystoreAPI().CreateUser(userPass)
+	if !ok || err != nil {
+		return fmt.Errorf("could not create user: %w", err)
+	}
+
+	txPChainClient := txClient.PChainAPI()
+	// Import genesis key
+	fundedAddress, err := txPChainClient.ImportKey(userPass, privateKey)
+	if err != nil {
+		return fmt.Errorf("unable to import genesis key: %w", err)
+	}
+	balance, err := txPChainClient.GetBalance(fundedAddress)
+	if err != nil {
+		return fmt.Errorf("unable to get genesis key balance: %w", err)
+	}
+	log.Info("found %d on address %s", balance.Balance, fundedAddress)
+
+	// Create a subnet
+	subnetIDTx, err := txPChainClient.CreateSubnet(
+		userPass,
+		[]string{fundedAddress},
+		fundedAddress,
+		[]string{fundedAddress},
+		defaultKeyThreshold,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create subnet: %w", err)
+	}
+
+CREATE_SUBNET:
```

Comment:
not very fan of labels in control flow. probably status get and check may be done outside of select .

Reply:
but we have to check for `ctx.Done()` ?

Reply:
I could extract these to individual funcs - let's wait for @danlaine 's review.

Reply:
I mean this (think should work)

```
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(loopTimeout):
		}
		status, _ := txPChainClient.GetTxStatus(subnetIDTx, true)
		if status.Status == platformvm.Committed {
			break
		}
		log.Debug("waiting for subnet creation tx (%s) to be accepted", subnetIDTx)
	}
	log.Info("subnet creation tx (%s) accepted", subnetIDTx)
```